torchbiggraph/storage_repository.py

```diff
@@ -0,0 +1,255 @@
+import logging
+import pathlib
+from abc import abstractmethod
+from contextlib import AbstractContextManager
+from io import TextIOWrapper
+from pathlib import Path
+from types import TracebackType
+from typing import IO
+from typing import Optional, Type, Union
+
+import h5py
+
+from torchbiggraph.plugin import PluginRegistry
+from torchbiggraph.util import url_path, run_external_cmd
+
+logger = logging.getLogger("torchbiggraph")
+
+
+class Constants:
+    GET = "hadoop fs -get {remote_path} {local_path}"
+    PUT = "hadoop fs -put -f {local_path} {remote_path}"
+    TEST_FILE = "hadoop fs -test -e {remote_path}"
+    REMOVE = "hadoop fs -rm {remote_path}"
+    H5 = "h5"
+    RELOAD = 'reload'
+    WRITE_MODES = ['w', 'x', 'a']
+    READ_MODE = 'r'
+
+
+class LocalFileContextManager(AbstractContextManager):
+
+    def __init__(self, path: Path, mode: str, **kwargs) -> None :
+        self._path: Path = path
+        self.mode = mode
+        self.kwargs = kwargs
+
+    def __enter__(self) -> Union[h5py.File, TextIOWrapper]:
+        self._file = self.get_resource(str(self._path), self.mode)
+        return self._file
+
+    def __exit__(self, exception_type: Optional[Type[BaseException]],
+                 exception_value: Optional[BaseException],
+                 traceback: Optional[TracebackType]) -> bool:
+        self._file.close()
+        self._file = None
+        return True
+
+    @staticmethod
+    def get_resource(filepath: str, mode: str) -> Union[h5py.File, IO[str]]:
+        # get file handler
+        if filepath.split(".")[-1] == Constants.H5:
+            # check if the file being read is a valid h5 file
+            if 'r' in mode:
+                if Path(filepath).exists() and not h5py.is_hdf5(filepath):
+                    raise ValueError('Invalid .h5 file', filepath)
+            return h5py.File(filepath, mode)
+        else:
+            return open(filepath, mode)
+
+
+class HDFSFileContextManager(AbstractContextManager):
+    def __init__(self, path: Path, mode: str, **kwargs) -> None :
+        self._path: Path = path
+        self.mode = mode
+        self.kwargs = kwargs
+
+    def __enter__(self) -> Union[h5py.File, TextIOWrapper]:
+        reload = True
+        if Constants.RELOAD in self.kwargs:
+            reload = self.kwargs[Constants.RELOAD]
+
+        if Constants.READ_MODE in self.mode:
+            self.get_from_hdfs(reload)
+
+        self._file = LocalFileContextManager.get_resource(str(self._path), self.mode)
+        return self._file
+
+
+    def __exit__(self, exception_type: Optional[Type[BaseException]],
+                 exception_value: Optional[BaseException],
+                 traceback: Optional[TracebackType]) -> bool:
+        self._file.close()
+        self._file = None
+
+        if any(ext in self.mode for ext in Constants.WRITE_MODES):
+            self.put_to_hdfs()
+
+        return True
+
+    @staticmethod
+    def get_hdfs_path(path) -> str:
```

Comment:
I'm a bit confused here... everything is mapped in HDFS and your cwd? This seems very messy...
- what if you have stale data in cwd?
- what if the data is too large to fit on local disk?